File: packages/pvm-assemblyscript/assembly/instructions/shifts-alt-64.ts
import {
  OPCODE_SHAR_R_IMM_ALT_64,
  OPCODE_SHLO_L_IMM_ALT_64,
  OPCODE_SHLO_R_IMM_ALT_64,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

export class SHLO_L_IMM_ALT_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_L_IMM_ALT_64
  name: string = 'SHLO_L_IMM_ALT_64'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper: reg'_A = (immed_X · 2^(reg_B mod 64)) mod 2^64
    // ALT: immediate << register (not register << immediate!)
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX

    const registerBValue = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = u64(registerBValue) % u64(64)

    // ALT: shift the immediate value by the register amount
    const result = u64(immediateX) << shiftAmount

    this.setRegisterValueWith64BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

export class SHLO_R_IMM_ALT_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_R_IMM_ALT_64
  name: string = 'SHLO_R_IMM_ALT_64'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper: reg'_A = floor(immed_X / 2^(reg_B mod 64))
    // ALT: immediate >> register (not register >> immediate!)
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX

    const registerBValue = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = u64(registerBValue) % u64(64)

    // ALT: shift the immediate value by the register amount (unsigned)
    const result = u64(immediateX) >> shiftAmount

    this.setRegisterValueWith64BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

export class SHAR_R_IMM_ALT_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHAR_R_IMM_ALT_64
  name: string = 'SHAR_R_IMM_ALT_64'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper: reg'_A = unsigned{floor(signed_64(immed_X) / 2^(reg_B mod 64))}
    // ALT: immediate >> register (not register >> immediate!) with arithmetic shift
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX

    const registerBValue = this.getRegisterValueAs64(context.registers, registerB)
    const shiftAmount = u64(registerBValue) % u64(64)

    // ALT: arithmetic shift the immediate value by the register amount
    const signedImmediate = this.toSigned64(immediateX)
    const shiftedValue = signedImmediate >> i64(shiftAmount)
    const result = this.toUnsigned64(shiftedValue)

    this.setRegisterValueWith64BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/shifts_alt_64.rs
//! Alternative shift by immediate (64-bit): immed_X << reg_B or immed_X >> reg_B.

use crate::config::{
    OPCODE_SHAR_R_IMM_ALT_64, OPCODE_SHLO_L_IMM_ALT_64, OPCODE_SHLO_R_IMM_ALT_64,
};
use crate::instructions::base::{
    arithmetic_shift_right_64, parse_two_registers_and_immediate, InstructionHandler,
};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register(registers: &mut [u64; 13], index: u8, value: u64) {
    if (index as usize) < 13 {
        registers[index as usize] = value;
    }
}

// SHLO_L_IMM_ALT_64: reg'_A = (immed_X << (reg_B mod 64)) mod 2^64
pub struct ShloLImmAlt64Instruction;
impl ShloLImmAlt64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloLImmAlt64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_L_IMM_ALT_64)
    }
    fn name(&self) -> &'static str {
        "SHLO_L_IMM_ALT_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let shift = get_register(context.registers, p.register_b) % 64;
        let imm = p.immediate_x as u64;
        let result = (imm << shift) & 0xffff_ffff_ffff_ffff;
        set_register(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHLO_R_IMM_ALT_64: reg'_A = floor(immed_X / 2^(reg_B mod 64))
pub struct ShloRImmAlt64Instruction;
impl ShloRImmAlt64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloRImmAlt64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_R_IMM_ALT_64)
    }
    fn name(&self) -> &'static str {
        "SHLO_R_IMM_ALT_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let shift = get_register(context.registers, p.register_b) % 64;
        let imm = p.immediate_x as u64;
        let result = imm >> shift;
        set_register(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHAR_R_IMM_ALT_64: reg'_A = unsigned{floor(signed_64(immed_X) / 2^(reg_B mod 64))}
pub struct SharRImmAlt64Instruction;
impl SharRImmAlt64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for SharRImmAlt64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHAR_R_IMM_ALT_64)
    }
    fn name(&self) -> &'static str {
        "SHAR_R_IMM_ALT_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let shift = (get_register(context.registers, p.register_b) % 64) as u32;
        let imm = p.immediate_x as u64;
        let result = arithmetic_shift_right_64(imm, shift);
        set_register(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
